homeassistant/components/smartthings/smartapp.py

```diff
@@ -0,0 +1,266 @@
+"""
+SmartApp functionality to receive cloud-push notifications.
+
+This module defines the functions to manage the SmartApp integration
+within the SmartThings ecosystem in order to receive real-time webhook-based
+callbacks when device states change.
+"""
+import asyncio
+import functools
+import logging
+import os
+from uuid import uuid4
+
+from aiohttp import web
+
+from homeassistant.components import webhook
+from homeassistant.const import CONF_WEBHOOK_ID
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.dispatcher import (
+    async_dispatcher_connect, async_dispatcher_send)
+from homeassistant.helpers.typing import HomeAssistantType
+from homeassistant.util.json import load_json, save_json
+
+from .const import (
+    APP_NAME_PREFIX, APP_OAUTH_SCOPES, CONF_APP_ID, CONF_INSTALLED_APP_ID,
+    CONF_INSTANCE_ID, CONF_LOCATION_ID, DATA_BROKERS, DATA_MANAGER, DOMAIN,
+    SETTINGS_INSTANCE_ID, SIGNAL_SMARTAPP_PREFIX, SMARTTHINGS_CONFIG_FILE,
+    SUPPORTED_CAPABILITIES)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def find_app(hass: HomeAssistantType, api):
+    """Find an existing SmartApp for this installation of hass."""
+    apps = await api.apps()
+    for app in [app for app in apps
+                if app.app_name.startswith(APP_NAME_PREFIX)]:
+        # Load settings to compare instance id
+        settings = await app.settings()
+        if settings.settings.get(SETTINGS_INSTANCE_ID) == \
+                hass.data[DOMAIN][CONF_INSTANCE_ID]:
+            return app
+
+
+async def validate_installed_app(api, installed_app_id: str):
+    """
+    Ensure the specified installed SmartApp is valid and functioning.
+
+    Query the API for the installed SmartApp and validate that it is tied to
+    the specified app_id and is in an authorized state.
+    """
+    from pysmartthings import InstalledAppStatus
+
+    installed_app = await api.installed_app(installed_app_id)
+    if installed_app.installed_app_status != InstalledAppStatus.AUTHORIZED:
+        raise RuntimeWarning("Installed SmartApp instance '{}' ({}) is not "
+                             "AUTHORIZED but instead {}."
+                             .format(installed_app.display_name,
+                                     installed_app.installed_app_id,
+                                     installed_app.installed_app_status))
+    return installed_app
+
+
+def _get_app_template(hass: HomeAssistantType):
+    from pysmartthings import APP_TYPE_WEBHOOK, CLASSIFICATION_AUTOMATION
+
+    return {
+        'app_name': APP_NAME_PREFIX + str(uuid4()),
+        'display_name': 'Home Assistant',
+        'description': "Home Assistant at " + hass.config.api.base_url,
```

Comment:
What about `hass.config.location_name` ?

Reply:
I figured `base_url` would more likely be unique than the `location_name` for those running multiple instances of HASS.  For example, If I have a dev HASS I build with separate from my "prod" instance, I probably haven't set different names -- but the base_url is going to be different.